### DISCUSSION TITLE
踩坑环境安装

### DISCUSSION THREAD

**Discussion by nimo12138** [InternLM/InternLM#127] [Show and tell]
# InternLM项目的依赖安装踩坑版

## 环境准备

- 实测能跑能训练
- 机器租的autodl 24g的3090 （非广，你也可以用本地或别家的）
- 机器环境：基础镜像 PyTorch 2.0.0 Python 3.8(ubuntu20.04) Cuda 11.8 自带conda3
- Python == 3.10（conda环境创建的3.10）
- GCC == 10.2.0（使用自带的9.4.0）
- MPFR == 4.1.0（使用自带的4.0.2）
- CUDA == 11.7（官网命令装的11.8）
- Pytorch == 1.13.1+cu117（官网命令装的2.0.0）
- Transformers >= 4.28.0（通过runtime.txt自动安装的4.29.2）
- Flash-Attention == v1.0.5
- Apex == 23.05
- Ampere或者Hopper架构的GPU (例如H100, A100)
- Linux OS

## 环境安装（需要学术加速，autodl帮助文档自带，其余环境自行想办法）

- 1. git clone https://github.com/InternLM/InternLM 下载项目
- 2. 进入 InternLM/third_party/ 文件夹
- 3. 删除 flash-attention 和 apex文件夹
- 4. 下载指定版本的 flash-attention 和 apex
        - 4.1 git clone https://github.com/NVIDIA/apex
        - 4.2 cd apex/
        - 4.3 git checkout 0da3ffb92ee6fbe5336602f0e3989db1cd16f880
        - 4.4 cd ..
        - 4.5 git clone https://github.com/Dao-AILab/flash-attention
        - 4.6 git checkout eff9fe6b8076df59d64d7a3f464696738a3c7c24
        - 4.7 cd ../..
- 5. 使用 conda 构建一个 Python-3.10 的虚拟环境
        - 5.1 conda create --name shusheng python=3.10 -y
        - 5.2 conda activate shusheng
- 6. 通过命令安装需要的依赖
        - 6.1 通过官网命令安装 pytorch 2.0.0和 cuda 11.8
            - conda install pytorch==2.0.0 torchvision==0.15.0 torchaudio==2.0.0 pytorch-cuda=11.8 -c pytorch -c nvidia
        - 6.2 安装运行需要的环境
            - pip install -r requirements/runtime.txt
            - 因为 6.1 命令装了pytorch和cuda，所以 “不要再执行 pip install -r requirements/torch.txt”
        - 6.3 安装 flash-attention (version v1.0.5)：
            - cd ./third_party/flash-attention
            - python setup.py install
            - cd ./csrc
            - cd fused_dense_lib && pip install -v .
            - cd ../xentropy && pip install -v .
            - cd ../rotary && pip install -v .
            - cd ../layer_norm && pip install -v .
            - cd ../../../../
        - 6.4 安装 Apex (version 23.05)：
            - cd ./third_party/apex
            - pip install -v --disable-pip-version-check --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./
            - cd ../../

- 7. 环境配好基本都能跑了
        - 7.1 修改 InternIM/internlm/solver/optimizer/hybrid_zero_optim.py" 中 “from torch._six import inf” 为 “from torch import inf” 不然微调和预训练会报错（pytorch 2.0.0包的区别）
        - 7.2 其余问题基本就是路径和配置文件参数的调整，修改参数后单卡24g的3090可以“重新训练”一个一言难尽的模型，微调需要的最低参数不清楚，但是8卡a100肯定没问题😅

**Comment by nimo12138** [comment]
代码忘标记可复制了😂懒得改

**Reply by vansin** [reply]
Thanks for your share~

**Comment by LiuZhihhxx** [comment]
你好，我在6.3安装flash-attention执行python setup.py install命令后，遇到CPU和内存跑到100%接着闪退的问题，偶尔还会shut掉浏览器进程。执行后终端显示内容如下：

ttorch.__version__  = 1.13.1+cu117


running install
/home/lz/anaconda3/envs/internlm-env/lib/python3.10/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
!!

        ********************************************************************************
        Please avoid running ``setup.py`` directly.
        Instead, use pypa/build, pypa/installer, pypa/build or
        other standards-based tools.

        See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
        ********************************************************************************

!!
  self.initialize_options()
/home/lz/anaconda3/envs/internlm-env/lib/python3.10/site-packages/setuptools/_distutils/cmd.py:66: EasyInstallDeprecationWarning: easy_install command is deprecated.
!!

        ********************************************************************************
        Please avoid running ``setup.py`` and ``easy_install``.
        Instead, use pypa/build, pypa/installer, pypa/build or
        other standards-based tools.

        See https://github.com/pypa/setuptools/issues/917 for details.
        ********************************************************************************

!!
  self.initialize_options()
running bdist_egg
running egg_info
writing flash_attn.egg-info/PKG-INFO
writing dependency_links to flash_attn.egg-info/dependency_links.txt
writing requirements to flash_attn.egg-info/requires.txt
writing top-level names to flash_attn.egg-info/top_level.txt
reading manifest file 'flash_attn.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*.cu' under directory 'flash_attn'
warning: no files found matching '*.h' under directory 'flash_attn'
warning: no files found matching '*.cuh' under directory 'flash_attn'
warning: no files found matching '*.cpp' under directory 'flash_attn'
warning: no files found matching '*.hpp' under directory 'flash_attn'
adding license file 'LICENSE'
adding license file 'AUTHORS'
writing manifest file 'flash_attn.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
running build_ext
building 'flash_attn_2_cuda' extensionorch.__version__  = 1.13.1+cu117


running install
/home/lz/anaconda3/envs/internlm-env/lib/python3.10/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
!!

        ********************************************************************************
        Please avoid running ``setup.py`` directly.
        Instead, use pypa/build, pypa/installer, pypa/build or
        other standards-based tools.

        See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
        ********************************************************************************

!!
  self.initialize_options()
/home/lz/anaconda3/envs/internlm-env/lib/python3.10/site-packages/setuptools/_distutils/cmd.py:66: EasyInstallDeprecationWarning: easy_install command is deprecated.
!!

        ********************************************************************************
        Please avoid running ``setup.py`` and ``easy_install``.
        Instead, use pypa/build, pypa/installer, pypa/build or
        other standards-based tools.

        See https://github.com/pypa/setuptools/issues/917 for details.
        ********************************************************************************

!!
  self.initialize_options()
running bdist_egg
running egg_info
writing flash_attn.egg-info/PKG-INFO
writing dependency_links to flash_attn.egg-info/dependency_links.txt
writing requirements to flash_attn.egg-info/requires.txt
writing top-level names to flash_attn.egg-info/top_level.txt
reading manifest file 'flash_attn.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*.cu' under directory 'flash_attn'
warning: no files found matching '*.h' under directory 'flash_attn'
warning: no files found matching '*.cuh' under directory 'flash_attn'
warning: no files found matching '*.cpp' under directory 'flash_attn'
warning: no files found matching '*.hpp' under directory 'flash_attn'
adding license file 'LICENSE'
adding license file 'AUTHORS'
writing manifest file 'flash_attn.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
running build_ext
/home/lz/anaconda3/envs/internlm-env/lib/python3.10/site-packages/torch/utils/cpp_extension.py:387: UserWarning: The detected CUDA version (11.5) has a minor version mismatch with the version that was used to compile PyTorch (11.7). Most likely this shouldn't be a problem.
  warnings.warn(CUDA_MISMATCH_WARN.format(cuda_str_version, torch.version.cuda))
building 'flash_attn_2_cuda' extension
Emitting ninja build file /home/lz/InternLM/third_party/flash-attention/build/temp.linux-x86_64-cpython-310/build.ninja...
Compiling objects...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
Emitting ninja build file /home/lz/InternLM/third_party/flash-attention/build/temp.linux-x86_64-cpython-310/build.ninja...
Compiling objects...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)


有关环境变量配置为：

declare -x CUDA_PATH="/usr/local/cuda"
declare -x GCC_HOME="/usr/bin/gcc"
declare -x MPFR_HOME="/usr/local/src/mpfr-4.2.0"
declare -x LD_LIBRARY_PATH="/usr/lib64:/usr/local/src/mpfr-4.2.0/lib:/usr/lib64:/usr/local/cuda/lib64"
declare -x PATH="/usr/bin:/usr/bin:/home/lz/anaconda3/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/cuda/bin"
declare -x CC="/usr/bin/gcc"
declare -x CXX="/usr/bin/c++"


请问是哪个环节出了问题？感激不尽！

**Reply by sunpengsdu** [reply]
根据https://github.com/Dao-AILab/flash-attention/issues/300 这里的一些说法，“I know PyTorch exposes the [MAX_JOBS environment variable](https://pytorch.org/docs/stable/cpp_extension.html#torch.utils.cpp_extension.BuildExtension) to prevent excessive parallel Ninja jobs from blowing up the system”， 需要设置MAX_JOBS这个环境变量，降低编译的并发量。Ninja会自动使用全部cpu，要是内部不够的话，就很容易oom

**Reply by nimo12138** [reply]
我这里用的pytorch是2.0.0的版本，看你错误提示里面的版本用的是1.13.1，你可以按照我用的版本依次安装试试，应该没有问题的，我之前用1.13.1也碰到很多问题没有全部解决，换了2.0.0就正常了

**Reply by LiuZhihhxx** [reply]
感谢 @sunpengsdu and @nimo12138 二位！
设置MAX_JOBS=n, n小于CPU数即可解决，保险起见我设置了1/2。
(pytorch版本仍为1.13.1)

**Comment by sunpengsdu** [comment]
https://hub.docker.com/r/sunpengsdu/internlm/tags

环境问题可以使用docker镜像，最新的main分支也修复了torch2.0使用的兼容问题

**Reply by tcexeexe** [reply]
你好，想问下在docker中如何启动web_demo，是否可以给一些提示？

**Comment by Wal-K-aWay** [comment]
你好，我在配环境的时候完全按照install.md的指示进行，在运行安装flash-attention (version v1.0.5)这一块时，输入 ../layer_norm && pip install -v .运行时会在[46/57]的地方卡住。
  [46/57] /usr/local/cuda/bin/nvcc  -I/data/users/weikewei/InternLM/third_party/flash-attention/csrc/layer_norm -I/data/users/weikewei/anaconda3/envs/InternLM-env/lib/python3.10/site-packages/torch/include -I/data/users/weikewei/anaconda3/envs/InternLM-env/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -I/data/users/weikewei/anaconda3/envs/InternLM-env/lib/python3.10/site-packages/torch/include/TH -I/data/users/weikewei/anaconda3/envs/InternLM-env/lib/python3.10/site-packages/torch/include/THC -I/usr/local/cuda/include -I/data/users/weikewei/anaconda3/envs/InternLM-env/include/python3.10 -c -c /data/users/weikewei/InternLM/third_party/flash-attention/csrc/layer_norm/ln_fwd_2560.cu -o /data/users/weikewei/InternLM/third_party/flash-attention/csrc/layer_norm/build/temp.linux-x86_64-cpython-310/ln_fwd_2560.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_BFLOAT16_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ -U__CUDA_NO_BFLOAT162_OPERATORS__ -U__CUDA_NO_BFLOAT162_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math -gencode arch=compute_70,code=sm_70 -gencode arch=compute_80,code=sm_80 --threads 4 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=dropout_layer_norm -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14
不知道是本来就会耗时比较长还是这里操作出错了？

**Reply by Wal-K-aWay** [reply]
Ctrl+C中断以后再次运行这条指令，就在Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)这里卡住了
